editor/config_loader.cpp

```diff
@@ -0,0 +1,176 @@
+{
+  HTTPClientPlatformWrapper request(url);
+  auto const reqResult = request.RunHTTPRequest();
+
+  if (reqResult &&
+      !request.was_redirected() &&
+      request.error_code() == 200)  // 200 - http status OK
+  {
+    return request.server_response();
+  }
+
+  return {};
+}
+}  // namespace
+
+namespace editor
+{
+void Waiter::Interrupt()
+{
+  {
+    lock_guard<mutex> lock(m_mutex);
+    m_interrupted = true;
+  }
+
+  m_event.notify_all();
+}
+
+ConfigLoader::ConfigLoader(pugi::xml_document & doc, mutex & mtx)
+  : m_docMutex(mtx)
+  , m_document(doc)
+{
+  LoadFromLocal();
+  m_loaderThr = thread(&ConfigLoader::LoadFromServer, this);
+}
+
+ConfigLoader::~ConfigLoader()
+{
+  m_waiter.Interrupt();
+
+  if (m_loaderThr.joinable())
+    m_loaderThr.join();
+}
+
+void ConfigLoader::LoadFromLocal()
+{
+  string content;
+  auto const reader = GetPlatform().GetReader(kConfigFileName);
+  reader->ReadAsString(content);
+
+  pugi::xml_document doc;
+  if (!doc.load_buffer(content.data(), content.size()))
+    MYTHROW(ConfigLoadError, ("Can't parse config"));
+
+  lock_guard<mutex> lock(m_docMutex);
+  m_document.reset(doc);
+}
+
+void ConfigLoader::LoadFromServer()
+{
+  auto const saveAndReload = [this](pugi::xml_document const & doc)
+  {
+    auto const filePath = GetConfigFilePath();
+    auto const result =
+      my::WriteToTempAndRenameToFile(filePath, [&doc](string const & fileName)
+      {
+        return doc.save_file(fileName.c_str());
+      });
+
+    if (!result)
+      return false;
+
+    LoadFromLocal();
+    return true;
+  };
+
+  auto const hash = LoadHash(GetHashFilePath());
+
+  try
+  {
+    do
+    {
+      auto const remoteHash = GetRemoteHash();
+      if (remoteHash.empty() || hash == remoteHash)
+       continue;
+
+      pugi::xml_document doc;
+      GetRemoteConfig(doc);
+
+      if (!doc.empty() && saveAndReload(doc))
+      {
+        SaveHash(hash, GetHashFilePath());
+      }
+    }
+    while(m_waiter.Wait(kSynchroTimeout));
+  }
+  catch (ConfigLoadError const & ex)
+  {
+    LOG(LERROR, (ex.Msg()));
+  }
+  catch (RootException const & ex)
+  {
+    LOG(LERROR, (ex.Msg()));
+  }
+}
+
+// static
+string ConfigLoader::GetRemoteHash()
+{
+  return RunSimpleHttpRequest(kRemoteHashUrl);
+}
+
+// static
+void ConfigLoader::GetRemoteConfig(pugi::xml_document &doc)
+{
+  auto const result = RunSimpleHttpRequest(kRemoteConfigUrl);
+  if (result.empty())
+    return;
+
+  if (!doc.load_string(result.c_str()))
+    doc.reset();
+}
+
+// static
+bool ConfigLoader::SaveHash(string const & hashStr, string const & filePath)
+{
+  auto const result =
+    my::WriteToTempAndRenameToFile(filePath, [&hashStr](string const & fileName)
+    {
+      ofstream ofs(fileName, ofstream::out);
+      if(!ofs.is_open())
+        return false;
+
+      ofs << hashStr;
+      return true;
+    });
+
+  return result;
+}
+
+// static
+string ConfigLoader::LoadHash(string const & filePath)
```

Comment:
You do not use any static fields, and the method is not used outside of the cpp. Please, move it to the anonymous namespace.